### PR TITLE
[ci]: test each individual crate's manifest

### PR DESCRIPTION
**Pull request by niklasad1**
We have been bitten by that the CI just checks all crates compiles in the workspace (shares Cargo.lock) and features.

Thus, if a feature for an individual is not enabled for specific crate then this is not detected by CI so this PR changes that each crate using its manifest instead.

In addition:
- Fixes the `ws-server` to compile outside the workspace
- Added to run CI on macos and Windows.

### CODE REVIEW

File: ws-server/src/tests.rs [hunk 1]
```diff
@@ -173,8 +173,6 @@ async fn can_set_max_connections() {
 	assert!(conn3.is_err());
 	let err = conn3.unwrap_err();
 	assert!(err.to_string().contains("WebSocketHandshake failed"));
-	assert!(err.to_string().contains("Connection reset by peer"));
-	// Err(Io(Os { code: 54, kind: ConnectionReset, message: \"Connection reset by peer\" }))");
```
Comment (niklasad1) [inline ws-server/src/tests.rs:177]: windows returns https://support.microsoft.com/en-us/topic/an-application-may-receive-the-10054-error-when-the-application-receives-data-from-a-connection-on-a-computer-that-is-running-windows-7-or-windows-server-2008-r2-if-a-tdi-filter-driver-is-installed-73fd74a9-51c2-3663-e965-e1187b1cfc92

so let's remove the assertion it would be better if we used `io::Error` in the test client I suppose

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -101,3 +161,59 @@ jobs:
...
+
+      - name: Install Rust stable toolchain
+        uses: actions-rs/toolchain@v1.0.7
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+
+      - name:                      Rust Cache
+        uses:                      Swatinem/rust-cache@v1.3.0
```
Comment (jsdw) [inline .github/workflows/ci.yml:180]: I'm curious; I've seen this spacing between the prop and value elsewhere for these cache lines and wasn't sure why; is there a reason for it?

Comment (niklasad1) [reply .github/workflows/ci.yml:180]: haha, we got a PR from the CI and he did this way but I can fix it ^

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -74,8 +74,68 @@ jobs:
           command: check
           args: --manifest-path http-client/Cargo.toml --no-default-features --features tokio02
```
Comment (jsdw) [inline .github/workflows/ci.yml:75]: Instead of checking every crate individually, could we just run `cargo check` without arguments? My experience is that that checks every crate in the workspace by default (though you can also add `--workspace` to, I assume, be sure).

Or do we need to check individually because `http-client` needs the `tokio02` feature setting?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -101,3 +161,59 @@ jobs:
...
+          override: true
+
+      - name:                      Rust Cache
+        uses:                      Swatinem/rust-cache@v1.3.0
+
+      - name: Cargo build
+        uses: actions-rs/cargo@v1.0.3
+        with:
+          command: build
+          args: --workspace
```
Comment (jsdw) [inline .github/workflows/ci.yml:186]: Do we need to have a build step before the `test` step? (I'll need to look into these rust github actions more!)

File: .github/workflows/ci.yml (hunk 2 above)
Comment (jsdw) [reply .github/workflows/ci.yml:180]: Aah; I think I removed the spaces when I copied some CI bits from jsonrpsee but I've been curious ever since :D Makes sense!

File: ws-server/Cargo.toml [hunk 5]
```diff
@@ -12,7 +12,7 @@ documentation = "https://docs.rs/jsonrpsee-ws-server"
 [dependencies]
 thiserror = "1"
 futures-channel = "0.3.14"
-futures-util = { version = "0.3.14", default-features = false, features = ["io"] }
+futures-util = { version = "0.3.14", default-features = false, features = ["io", "async-await-macro"] }
```
Comment (jsdw) [inline ws-server/Cargo.toml:15]: Nit: It doesn't look like we've added any code that requires this feature; do we want it?

**Review by jsdw — APPROVED** [review]
Looks good to me; I have a few questions but nothing major and mainly just for my own curiosity!

File: ws-server/Cargo.toml (hunk 5 above)
Comment (niklasad1) [reply ws-server/Cargo.toml:15]: it was introduced in another PR, thus doesn't compile standalone such as if you `cd ws-server && cargo check` outside the workspace it doesn't work. Because, features in Rust are additive (the union of all features) so if another crate adds that feature it's enabled for all in workspace if that makes sense.

TLDR; on master `cd ws-server && cargo check` doesn't work but `cargo build -p ws-server` works :P

File: .github/workflows/ci.yml (hunk 3 above)
Comment (niklasad1) [reply .github/workflows/ci.yml:75]: yeah, we did do `cargo check --workspace` before this PR but that's the problem and why I added `cargo check --manifest-path crate`, see my comment below.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (niklasad1) [reply .github/workflows/ci.yml:186]: no you are right, we don't have too but the previous checks only run `cargo check` so we have want make sure nothing weird happens when building the binary but it could be another action for clarity I suppose.

File: ws-server/Cargo.toml (hunk 5 above)
Comment (jsdw) [reply ws-server/Cargo.toml:15]: Aah gotcha! Thanks for the explanation!

File: .github/workflows/ci.yml (hunk 4 above)
Comment (jsdw) [reply .github/workflows/ci.yml:186]: My first thought was that `cargo test` builds the binary anyway, but of course it builds it with `cfg(test)` enabled, so I can see why a separate build step is useful to make sure it works "normally"!

File: .github/workflows/ci.yml (hunk 3 above)
Comment (jsdw) [reply .github/workflows/ci.yml:75]: Thanks, that all makes sense now :)

File: ws-server/src/tests.rs [hunk 6]
```diff
@@ -171,10 +171,12 @@ async fn can_set_max_connections() {
...
-	let err = conn3.unwrap_err();
-	assert!(err.to_string().contains("WebSocketHandshake failed"));
-	assert!(err.to_string().contains("Connection reset by peer"));
-	// Err(Io(Os { code: 54, kind: ConnectionReset, message: \"Connection reset by peer\" }))");
+
+	let err = match conn3 {
+		Err(soketto::handshake::Error::Io(err)) => err,
+		_ => panic!("Invalid error kind; expected std::io::Error"),
+	};
+	assert_eq!(err.kind(), std::io::ErrorKind::ConnectionReset);
```
Comment (dvdplm) [inline ws-server/src/tests.rs:179]: Nice, ty for fixing this.

**Review by dvdplm — APPROVED** [review]
